x-pack/plugins/enterprise_search/public/applications/workplace_search/views/content_sources/components/add_source/add_source_logic.ts

```diff
@@ -492,12 +492,18 @@ export const AddSourceLogic = kea<MakeLogicType<AddSourceValues, AddSourceAction
 
       try {
         const response = await http.get(route, { query });
+        const { serviceName, indexPermissions, serviceType, preContentSourceId } = response;
 
-        const { serviceName, indexPermissions, serviceType } = response;
+        // GitHub requires an intermediate configuration step, where we collect the repos to index.
+        if (preContentSourceId) {
+          actions.setPreContentSourceId(preContentSourceId);
+          navigateToUrl(`${ADD_GITHUB_PATH}/configure${search}`);
+        } else {
         setAddedSource(serviceName, indexPermissions, serviceType);
+          navigateToUrl(getSourcesPath(SOURCES_PATH, isOrganization));
+        }
       } catch (e) {
         flashAPIErrors(e);
-      } finally {
         navigateToUrl(getSourcesPath(SOURCES_PATH, isOrganization));
```

Comment:
Could we keep the finally block and navigate to correct URL inside it based on value inside `if`?

Something like this:
```javascript
finally {
  if (hasConfigureStep && !values.oauthConfigCompleted) {
    navigateToUrl(`${ADD_GITHUB_PATH}/configure${search}`);
  } else {
    navigateToUrl(getSourcesPath(SOURCES_PATH, isOrganization));
  }
}
```

That way we have less duplication and the intention is clearer.

Reply:
I started to do that but the `hasConfigureStep` is not in the scope of the `finally` block; it is returned from the HTTP response and declaring a variable outside of scope for this just didn't feel right. 

![image](https://user-images.githubusercontent.com/1869731/112505362-b5b45700-8d5a-11eb-8b8b-dca4c0723906.png)

We would also still need 2 conditionals in the method you are proposing because we still need the `if` block to call the `actions.setPreContentSourceId` method in the `try` block. 

Happy to discuss further and tidy it up in a separate PR.

Reply:
Thanks for explaining, I think it's fine as-is.

Reply:
I appreciate the feedback and think this could be done a lot better as a whole.